File: aznavrail-react/src/__tests__/AzGuidanceExpansion.test.ts
import { routeInstructions } from '../guidance/AzGuidance';
import {
  AZ_ITEM_ACTIVE,
  resolveAzHighlight,
  resolveItemId,
  resolveShape,
  shapeBounds,
} from '../guidance/AzStatus';
import { anySuppressorActive } from '../guidance/AzStatusEngine';
import type { AzEdge, AzGoal, AzGuideShape, AzItemBounds } from '../guidance/AzStatus';

/** Parity with the Kotlin AzGuidanceExpansionTest: paged steps, highlight resolution, suppression. */

const stepped: AzEdge = {
  from: 'a',
  to: 'done',
  instruction: { text: 'step0' },
  steps: [
    { text: 'step0' },
    { text: 'step1', highlightTargetId: 'ball', advanceWhen: 'x' },
    { text: 'step2' },
  ],
};
const goals: Record<string, AzGoal> = { g: { id: 'g', target: 'done' } };

describe('paged routing', () => {
  it('shows the step at the cursor', () => {
    const f0 = routeInstructions([stepped], goals, ['g'], new Set(['a']), () => 0);
    expect(f0.instructions[0].text).toBe('step0');
    expect(f0.resolved[0].stepIndex).toBe(0);
    expect(f0.resolved[0].stepTotal).toBe(3);

    const f1 = routeInstructions([stepped], goals, ['g'], new Set(['a']), () => 1);
    expect(f1.instructions[0].text).toBe('step1');
    expect(f1.resolved[0].instruction.highlight).toEqual({ type: 'Target', id: 'ball' });
  });

  it('reactive advanceWhen wins over the tap cursor', () => {
    const f = routeInstructions([stepped], goals, ['g'], new Set(['a', 'x']), () => 1);
    expect(f.instructions[0].text).toBe('step2');
    expect(f.resolved[0].stepIndex).toBe(2);
  });

  it('does not skip an informational step when a later step status is already true', () => {
    const f = routeInstructions([stepped], goals, ['g'], new Set(['a', 'x']), () => 0);
    expect(f.instructions[0].text).toBe('step0');
  });

  it('reaches a paged goal when its target becomes true', () => {
    const f = routeInstructions([stepped], goals, ['g'], new Set(['a', 'done']), () => 1);
    expect(f.reachedGoals.has('g')).toBe(true);
    expect(f.instructions).toHaveLength(0);
  });

  it('keeps single-callout edges unchanged (stepTotal 1)', () => {
    const edge: AzEdge = { from: 'a', to: 'b', instruction: { text: 'Do A', highlight: { type: 'Item', id: 'ia' } } };
    const f = routeInstructions([edge], { g: { id: 'g', target: 'b' } }, ['g'], new Set(['a']));
    expect(f.instructions[0].text).toBe('Do A');
    expect(f.resolved[0].stepTotal).toBe(1);
  });
});

describe('highlight + shape resolution', () => {
  it('resolveAzHighlight precedence is target, selector, active token, item, none', () => {
    expect(resolveAzHighlight('item', () => 'sel', 't')).toEqual({ type: 'Target', id: 't' });
    expect(resolveAzHighlight('item', () => 'sel', undefined).type).toBe('Dynamic');
    expect(resolveAzHighlight(AZ_ITEM_ACTIVE, undefined, undefined)).toEqual({ type: 'ActiveItem' });
    expect(resolveAzHighlight('item', undefined, undefined)).toEqual({ type: 'Item', id: 'item' });
    expect(resolveAzHighlight(undefined, undefined, undefined)).toEqual({ type: 'None' });
  });

  it('resolveItemId folds ActiveItem and Dynamic against the live active id', () => {
    expect(resolveItemId({ type: 'ActiveItem' }, 'home')).toBe('home');
    expect(resolveItemId({ type: 'Dynamic', selector: () => 'layer.1' }, null)).toBe('layer.1');
    expect(resolveItemId({ type: 'Dynamic', selector: () => AZ_ITEM_ACTIVE }, 'home')).toBe('home');
    expect(resolveItemId({ type: 'Target', id: 't' }, 'home')).toBeNull();
  });

  it('resolveShape resolves targets, areas and items, degrading to null', () => {
    const circle: AzGuideShape = { type: 'Circle', cx: 1, cy: 2, radius: 3 };
    const targets = { t: () => circle, gone: () => null };
    expect(resolveShape({ type: 'Target', id: 't' }, {}, null, targets)).toEqual(circle);
    expect(resolveShape({ type: 'Target', id: 'gone' }, {}, null, targets)).toBeNull();
    expect(resolveShape({ type: 'Target', id: 'missing' }, {}, null, targets)).toBeNull();
    expect(resolveShape({ type: 'Area', left: 5, top: 6, width: 7, height: 8 }, {}, null, {})?.type).toBe('Rect');

    const cache: Record<string, AzItemBounds> = { home: { x: 0, y: 0, width: 10, height: 10 } };
    expect(resolveShape({ type: 'Item', id: 'home' }, cache, null, {})?.type).toBe('Rect');
    expect(resolveShape({ type: 'Item', id: 'nope' }, cache, null, {})).toBeNull();
    expect(resolveShape({ type: 'ActiveItem' }, cache, 'home', {})?.type).toBe('Rect');
    expect(resolveShape({ type: 'ActiveItem' }, cache, null, {})).toBeNull();
  });

  it('shapeBounds wraps each geometry with padding', () => {
    expect(shapeBounds({ type: 'Circle', cx: 100, cy: 50, radius: 10, padding: 5 })).toEqual({ left: 85, top: 35, width: 30, height: 30 });
    expect(shapeBounds({ type: 'Rect', left: 10, top: 20, width: 30, height: 40 })).toEqual({ left: 10, top: 20, width: 30, height: 40 });
    expect(
      shapeBounds({ type: 'Path', commands: [{ type: 'M', x: 0, y: 0 }, { type: 'Q', x1: 5, y1: 30, x: 10, y: 20 }, { type: 'Z' }] }),
    ).toEqual({ left: 0, top: 0, width: 10, height: 30 });
  });
});

describe('suppression', () => {
  it('anySuppressorActive ORs predicates and treats a throw as false', () => {
    expect(anySuppressorActive([])).toBe(false);
    expect(anySuppressorActive([[700, () => true]])).toBe(true);
    expect(anySuppressorActive([[700, () => false]])).toBe(false);
    expect(anySuppressorActive([[0, () => false], [0, () => true]])).toBe(true);
    expect(anySuppressorActive([[0, () => { throw new Error('boom'); }]])).toBe(false);
  });
});

File: aznavrail/src/test/java/com/hereliesaz/aznavrail/tutorial/AzGuidanceExpansionTest.kt
package com.hereliesaz.aznavrail.tutorial

import androidx.compose.ui.geometry.Rect
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Covers the guidance-framework expansion: paged steps + reactive/manual advance, arbitrary-shape
 * highlight resolution, the highlight-directive precedence, and the suppression predicate combiner.
 */
class AzGuidanceExpansionTest {

    // a --(3 paged steps)--> done
    private val stepped = AzEdge(
        from = "a",
        to = "done",
        instruction = AzInstruction("step0"),
        steps = listOf(
            AzInstructionStep("step0"),
            AzInstructionStep("step1", highlightTargetId = "ball", advanceWhen = "x"),
            AzInstructionStep("step2"),
        ),
    )
    private val goals = mapOf("g" to AzGoal(id = "g", target = "done"))

    @Test
    fun `paged edge shows the step at the cursor`() {
        val f0 = routeInstructions(listOf(stepped), goals, listOf("g"), setOf("a"), stepIndexOf = { 0 })
        assertEquals("step0", f0.instructions.single().text)
        assertEquals(0, f0.resolved.single().stepIndex)
        assertEquals(3, f0.resolved.single().stepTotal)

        val f1 = routeInstructions(listOf(stepped), goals, listOf("g"), setOf("a"), stepIndexOf = { 1 })
        assertEquals("step1", f1.instructions.single().text)
        assertEquals(AzGuideHighlight.Target("ball"), f1.resolved.single().instruction.highlight)
    }

    @Test
    fun `reactive advanceWhen wins over the tap cursor`() {
        // Cursor parked on step1, whose advanceWhen ("x") is satisfied -> the overlay shows step2.
        val f = routeInstructions(listOf(stepped), goals, listOf("g"), setOf("a", "x"), stepIndexOf = { 1 })
        assertEquals("step2", f.instructions.single().text)
        assertEquals(2, f.resolved.single().stepIndex)
    }

    @Test
    fun `an informational step is not skipped when a later step's status is already true`() {
        // step0 has no advanceWhen, so it stays even though "x" (step1's trigger) is true.
        val f = routeInstructions(listOf(stepped), goals, listOf("g"), setOf("a", "x"), stepIndexOf = { 0 })
        assertEquals("step0", f.instructions.single().text)
    }

    @Test
    fun `a paged goal is reached when its target status becomes true`() {
        val f = routeInstructions(listOf(stepped), goals, listOf("g"), setOf("a", "done"), stepIndexOf = { 1 })
        assertTrue("g" in f.reachedGoals)
        assertTrue(f.instructions.isEmpty())
    }

    @Test
    fun `single-callout edges still route unchanged (stepTotal 1)`() {
        val edge = AzEdge("a", "b", AzInstruction("Do A", highlight = AzGuideHighlight.Item("ia")))
        val g = mapOf("g" to AzGoal("g", "b"))
        val f = routeInstructions(listOf(edge), g, listOf("g"), setOf("a"))
        assertEquals("Do A", f.instructions.single().text)
        assertEquals(1, f.resolved.single().stepTotal)
    }

    @Test
    fun `resolveAzHighlight precedence is target, selector, active token, item, none`() {
        assertEquals(AzGuideHighlight.Target("t"), resolveAzHighlight("item", { "sel" }, "t"))
        assertTrue(resolveAzHighlight("item", { "sel" }, null) is AzGuideHighlight.Dynamic)
        assertEquals(AzGuideHighlight.ActiveItem, resolveAzHighlight(AZ_ITEM_ACTIVE, null, null))
        assertEquals(AzGuideHighlight.Item("item"), resolveAzHighlight("item", null, null))
        assertEquals(AzGuideHighlight.None, resolveAzHighlight(null, null, null))
    }

    @Test
    fun `resolveItemId folds ActiveItem and Dynamic against the live active id`() {
        assertEquals("home", AzGuideHighlight.ActiveItem.resolveItemId("home"))
        assertEquals("layer.1", AzGuideHighlight.Dynamic { "layer.1" }.resolveItemId(null))
        assertEquals("home", AzGuideHighlight.Dynamic { AZ_ITEM_ACTIVE }.resolveItemId("home"))
        assertNull(AzGuideHighlight.Target("t").resolveItemId("home"))
    }

    @Test
    fun `resolveShape resolves targets, areas and items, degrading to null`() {
        val circle = AzGuideShape.Circle(1f, 2f, 3f)
        val targets = mapOf<String, () -> AzGuideShape?>("t" to { circle }, "gone" to { null })
        assertEquals(circle, AzGuideHighlight.Target("t").resolveShape(emptyMap(), null, targets))
        assertNull(AzGuideHighlight.Target("gone").resolveShape(emptyMap(), null, targets))
        assertNull(AzGuideHighlight.Target("missing").resolveShape(emptyMap(), null, targets))
        assertTrue(AzGuideHighlight.Area(5f, 6f, 7f, 8f).resolveShape(emptyMap(), null, emptyMap()) is AzGuideShape.Rect)

        val cache = mapOf("home" to Rect(0f, 0f, 10f, 10f))
        assertTrue(AzGuideHighlight.Item("home").resolveShape(cache, null, emptyMap()) is AzGuideShape.Rect)
        assertNull(AzGuideHighlight.Item("nope").resolveShape(cache, null, emptyMap()))
        assertTrue(AzGuideHighlight.ActiveItem.resolveShape(cache, "home", emptyMap()) is AzGuideShape.Rect)
        assertNull(AzGuideHighlight.ActiveItem.resolveShape(cache, null, emptyMap()))
    }

    @Test
    fun `AzGuideShape bounds wraps each geometry with padding`() {
        val c = AzGuideShape.Circle(100f, 50f, 10f, padding = 5f).bounds()
        assertEquals(Rect(85f, 35f, 115f, 65f), c)
        val r = AzGuideShape.Rect(10f, 20f, 30f, 40f).bounds()
        assertEquals(Rect(10f, 20f, 40f, 60f), r)
        val p = AzGuideShape.Path(
            listOf(AzPathCmd.MoveTo(0f, 0f), AzPathCmd.QuadTo(5f, 30f, 10f, 20f), AzPathCmd.Close),
        ).bounds()
        assertEquals(Rect(0f, 0f, 10f, 30f), p)
    }

    @Test
    fun `anySuppressorActive ORs predicates and treats a throw as false`() {
        assertFalse(anySuppressorActive(emptyList()))
        assertTrue(anySuppressorActive(listOf(700L to { true })))
        assertFalse(anySuppressorActive(listOf(700L to { false })))
        assertTrue(anySuppressorActive(listOf(0L to { false }, 0L to { true })))
        assertFalse(anySuppressorActive(listOf(0L to { throw RuntimeException("boom") })))
    }
}
